### ISSUE TITLE
macOS unable to compile with openCL support

### ISSUE THREAD

**Issue by cedricloneux** [ImageMagick/ImageMagick#5614]
### ImageMagick version

7.1.0

### Operating system

MacOS

### Operating system, version and so on

13 (ventura)

### Description

Trying to compile with OpenCL supports on macOS.

I tried to compile manually but got an error. I tried via home-brew. open-cl is disabled but default in this one so I did
```brew edit imagemagick``` and removed the _--disable-opencl_ arg and added _--enable-opencl_
it worked as I got the same error as when building from source

```
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [MagickCore/libMagickCore-7.Q16HDRI.la] Error 1
```

### from source:
./configure --enable-opencl
make

### homebrew
brew edit imagemagick
remove --disable-opencl flag
add --enable-opencl flag
brew reinstall -s imagemagick

**Comment by fmw42** [comment]
I am not an expert on this, but you need to have the proper graphic card.

**Comment by cedricloneux** [comment]
what do you mean ? MacBooks have a proper internal graphic card that. I use openCL in many other softwares

**Comment by fmw42** [comment]
All I meant was that it has to be compatible with OpenCL.

**Comment by rlynch-bitpusher** [comment]
I have an `22 M1 MBpro with ventura and can corroborate this course of action results in the above error.  I have even gone so far as to brew install the opencl-headers and export the headerpath into the environment - without success.

**Comment by lmoberly** [comment]
This posting describes the same issue I have, using macOS Monterey. There are more than a dozen OpenCL related reference errors [1] in the configuration code that give the "clang: error: linker command failed" error at the end of the make step. (all new to me so I don't know the terminology). I am able to install ImageMagic both from source and with Homebrew, but, not with the  "--enable-opencl" option. Has anyone seen an understandable way to fix this?

[1] Undefined symbols for architecture x86_64:
  "_clBuildProgram", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clCreateBuffer", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clCreateCommandQueue", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clCreateContext", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o ...............

**Comment by fmw42** [comment]
My guess is that Home-brew uses a compiler that is not compatible with compiling in OpenCL.  Perhaps one of the IM developers can confirm.

**Comment by lmoberly** [comment]
But we are seeing pretty much the same error building magick from source:

> ./configure --enable-opencl
> make

I can't express it but it looks like the Image Magick use of OpenCL is using code names ("_clBuildProgram" ??) that it can't find in the Apple macOS libraries, or something along that line.

**Comment by fmw42** [comment]
I can confirm that I get errors when trying to manually "make" the install on Mac OSX Monterey on M1 Mac trying to build IM 7.1.0.53 with OpenCL.  (My delegates are installed using MacPorts, but I build IM from source). It works fine without OpenCL.

Here are my messages at the make stage:

```
Undefined symbols for architecture arm64:
  "_clBuildProgram", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clCreateBuffer", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clCreateCommandQueue", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clCreateContext", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clCreateKernel", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clCreateProgramWithBinary", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clCreateProgramWithSource", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clEnqueueMapBuffer", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clEnqueueNDRangeKernel", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clEnqueueReadBuffer", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clEnqueueUnmapMemObject", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clFinish", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clFlush", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clGetDeviceIDs", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clGetDeviceInfo", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clGetEventInfo", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clGetEventProfilingInfo", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clGetKernelInfo", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clGetPlatformIDs", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clGetPlatformInfo", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clGetProgramBuildInfo", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clGetProgramInfo", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clReleaseCommandQueue", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clReleaseContext", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clReleaseEvent", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clReleaseKernel", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clReleaseMemObject", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clReleaseProgram", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clRetainEvent", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clRetainMemObject", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clSetEventCallback", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clSetKernelArg", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
  "_clWaitForEvents", referenced from:
      _InitializeOpenCL in libMagickCore_7_Q16HDRI_la-opencl.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [MagickCore/libMagickCore-7.Q16HDRI.la] Error 1
make: *** [all] Error 2
```

Here is my configure command:

```
./configure \
--enable-openmp  \
--enable-delegate-build --enable-shared --disable-static --enable-opencl \
--with-modules --with-quantum-depth=16 --without-wmf --with-rsvg --with-fftw \
--disable-silent-rules --disable-dependency-tracking --without-pango \
--with-lqr --with-gslib --with-gs-font-dir=/opt/local/share/ghostscript/fonts/

```

**Comment by lmoberly** [comment]
OpenMP(5.0)  is s feature in my installed IM (without the OpenCL) but I see these same OpenCL errors when I try to configure with OpenCL. Web references to IM OpenCL are scattered and mostly not related to macOS. I had not heard of IM before two weeks ago. I also just bought a new video module and am curious just what I can get the gpu to do. Maybe we'll get a few replies that will help with the openCL stuff.

**Comment by lmoberly** [comment]
There are many google responses for "Undefined symbols for architecture x86_64: "_clBuildProgram", referenced from:.....: "  
Like this one: https://www.positioniseverything.net/undefined-symbols-for-architecture-x86_64

They are all related to how Xcode builds different OS programs, but very similar type errors.  So far I have not understood any of these references enough to apply to IM.

**Comment by dlemstra** [comment]
I noticed this in your build log: `ld: symbol(s) not found for architecture arm64`. It looks like you are mixing assemblies of different build types? Looks like your OpenCL library is build with x64?

**Comment by cedricloneux** [comment]
@dlemstra why closing ? We are still all having the issue

**Comment by dlemstra** [comment]
There has been no response to my last message that I posted in the beginning of this year. If we don't get answers we will at some point close the issue to reduce the number of open issues that we need to take a look at.

**Comment by mohd-akram** [comment]
Running into this as well, kindly reopen. I'm on an Intel Mac. During the configure stage, it can find the OpenCL framework:

```
checking for OpenCL library... checking for framework OpenCL... yes
yes
```

But it does not seem to be persisted in `LIBS`. Looking at `ax_have_opencl.m4` it seems to be added to `LIBS_CL` but that isn't used anywhere.

**Comment by mohd-akram** [comment]
@dlemstra This is the exact same issue, I get the same undefined symbols. The fix seems to be to rename `LIBS_CL` to `CL_LIBS` in `ax_have_opencl.m4`. `CL_LIBS` is used in `configure.ac`. This is a potential patch:

```diff
--- m4/ax_have_opencl.m4.orig	2023-09-23 17:17:14
+++ m4/ax_have_opencl.m4	2023-09-27 19:49:24
@@ -87,12 +87,12 @@
       #
       # First we check for Mac OS X, since OpenCL is standard there
       #
-      LIBS_CL="none"
+      CL_LIBS=""
       AC_MSG_CHECKING([for OpenCL library])
       case "$host_os" in
         darwin*) # On Mac OS X we check for installed frameworks
         AX_CHECK_FRAMEWORK([OpenCL], [
-          LIBS_CL="-framework OpenCL"
+          CL_LIBS="-framework OpenCL"
           no_cl=no
           AC_MSG_RESULT(yes)
           CL_ENABLED=true
@@ -126,7 +126,7 @@
           AC_MSG_RESULT(yes)
           CL_ENABLED=true
           CL_VERSION=1
-          LIBS_CL="$save_LIBS -L/usr/lib64/nvidia -L/usr/lib/nvidia -lOpenCL"
+          CL_LIBS="-L/usr/lib64/nvidia -L/usr/lib/nvidia -lOpenCL"
         ],[
           no_cl=yes
           AC_MSG_RESULT(no)
```

**Comment by dlemstra** [comment]
From the information that you provided that was not possible to deduct that you ran into the same issue. This does look like the correct patch. Can you open a PR add this path?

**Comment by cedricloneux** [comment]
I still can't use GPU on mac.
I have tried again with version 7.1.1-21.

First try:
`brew edit imagemagick` and removed the --disable-opencl arg and added --enable-opencl. Then `brew install --build-from-source imagemagick` But at the configuring step it still print _./configure ... -disable-opencl_ why ?

Second try:
So I tried to download the last release as source code and compile myself. `.configure --enable-opencl` suceed then `sudo make` which result in a error:

```
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [MagickCore/libMagickCore-7.Q16HDRI.la] Error 1
```

Btw Apple has deprecated OpenCL since years now and even this one doesn't seems to work on mac. I suppose Apple will eventually remove openCL support completely. Will imagemagick be Metal compatible ?
